text/0000-ubi-based-stacks.md

```diff
@@ -0,0 +1,181 @@
+
+
+## Rationale and Alternatives
+
+With the goal in mind of creating a Stack based on UBI using system packaged depdencies, these alternatives 
+have been considered...
+
+#### Builders per Runtime
+This option would have 1 Builder image per runtime, paired with a matching run image per runtime, with the runtime
+depdencies baked into the builder/run images. This option did not require Extensions from upstream, but would 
+result in a large collection of single purpose stacks which rapidly becomes unmanageable as you consider versioning
+across runtimes (eg, there would likely need to be java 8,11,17 stacks). Exisiting Paketo buildpacks would be 
+checked to ensure they would honor & use a dependency already in the builder image, over one the buildpack would
+normally download & use.
+
+In addition this transfers the responsibility of a lot of the 'magic' of Buildpacks to the end user, requiring them 
+to know in advance which builder image to use with a given project, and to submit to the correct build pipeline, or
+adapt the pipeline per project. 
+
+#### 'Phat' Builder Image
+This option would have 1 Builder image with many runtimes baked in, and then a selection of run images with
+appropriate runtimes baked in that would be selected by the Builder via Extensions. As with the previous option, 
+existing buildpacks would be checked to ensure they would use the baked in dependencies as a priority. 
+
+This approach leads to a very large Builder image, which is suboptimal from an end user perspective, plus, as all 
+runtimes are baked in, there would likely be frequent updates to the Builder image to incorporate fixes & security
+updates to each contained runtime, leading to the end-user requiring a fresh version of the builder image for almost 
+every update. A large image + frequent updates creates a less than ideal development experience.
+
+#### UBI without rpms
+This option would just take the existing Paketo buildpacks as-is and run them on a UBI base image. As with all the 
+options here, testing would be required for all contained buildpacks to ensure any system or shell interactions were 
+unaffected by the Ubuntu->UBI switch. 
+
+Although this approach would produce functional application images, the resulting images would not be using UBI 
+installed runtimes, and thus could be unable to qualify for full support in production. This is considered important 
+enough to end users that this option was discarded.
+
+### Chosen approach.
+The chosen approach of using an empty Builder image with Extension installed depdencies via RPM meets the goals of 
+having a lightweight development experience, while still being able to create supportable applications based on UBI 
+images. 
+
+## Implementation
+
+### Repositories
+
+The stacks team will create a new repo for UBI.. 
+
+* `ubi-base-stack`
+
+The buildpacks team will create two new repos for the initial two Extensions.
+
+* `ubi-java-extension`
+* `ubi-nodejs-extension`
```

Comment:
Are you proposing these be owned by each language family team? i.e. Java buildpacks team owns `ubi-java-extension` and `ubi-nodejs-extension` is owned by the Node.js team?

Do you have a link to these you can share? I'm not sure if it's part of this RFC, possibly another one, or just an implementation detail, but these are going to be the first extensions we support so I think it would be worth digging into the way these are set up, what tooling is being used, and what sorts of patterns we're trying to establish for future extensions.

Reply:
To your point Dan, I'm hesitant about adding specifics about the particular tooling and patterns in this RFC. My biggest concern is that we'd pigeonhole ourselves into a specific implementation across the board, which feels especially unnecessary if each extension will be maintained by separate language maintainer groups. It doesn't hurt to align on these things, but I'm not sure if declaring these things at the RFC level is too much.

Reply:
I should add that just because it makes sense for these two extensions to be owned by their respective language groups doesn't mean that will be true for all extensions.. eg an 'update-ca-certificates' extension (that updates the system certs in the builder image during build to include self-signed certs required for local resources).. wouldn't have any affinity to either java, or node, or any specific runtime. 

It's also maybe worth pointing out that it's doable to have multiple extensions for a given language group, eg Java could have a ubi-maven-extension that 'supplied' maven via rpm (in practice this isn't required because the maven buildpack can defer to an existing on-path maven, but you get the idea). 

The only critical thing about use of extensions (enforced by the human who selects which extensions are packaged into the builder image) is that they don't end up fighting over which run image they want to switch to. This is fine with a hypothetical ca-certificates extension that doesn't want to touch the run image, and just needs to fix up the build env, but less good if say node & java both decide to kick in for the same project. (we're aware of the recent java/node changes with buildpacks)

Reply:
@sophiewigmore Yes, I agree. I don't want to make the RFC more complicated or overly specific. I do have concerns with having consistent implementations though, not at a fine-grained level, but in a larger sense. In short, I want to try to avoid another libpak vs packit situation.